### PR TITLE
Adds ability for crawlers to write to other catalogs than hive_metastore

### PR DESCRIPTION
**Pull request by pritishpai**
## Changes
Currently, all crawlers write to the inventory database under the hard-coded "hive_metastore". This will allow the users to use a custom catalog for writing ucx metadata.

### Linked issues
Introduces #2575 

### Functionality

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests

- [ ] manually tested
- [ ] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: src/databricks/labs/ucx/account/aggregate.py [hunk 1]
```diff
@@ -75,12 +75,13 @@ def _federated_ucx_query(self, query: str, table_name='objects') -> Iterable[tup
...
                             src_schema=inventory_database,
                             src_table=table_name,
-                            dst_catalog='hive_metastore',
+                            dst_catalog=ctx.config.inventory_catalog,
                             dst_schema=inventory_database,
                             dst_table=table_name,
                         )
                     ]
                 )
+                # TODO: check whether to add catalog here as well
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/account/aggregate.py:84]: I expect not

File: src/databricks/labs/ucx/contexts/application.py [hunk 2]
```diff
@@ -278,15 +284,17 @@ def table_move(self):
...
 
     @cached_property
     def azure_service_principal_crawler(self):
         return AzureServicePrincipalCrawler(self.workspace_client, self.sql_backend, self.inventory_database)
 
     @cached_property
     def external_locations(self):
-        return ExternalLocations(self.workspace_client, self.sql_backend, self.inventory_database)
+        return ExternalLocations(
+            self.workspace_client, self.sql_backend, self.inventory_catalog, self.inventory_database
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/contexts/application.py:296]:
```suggestion
            self.workspace_client, self.sql_backend, self.inventory_catalog, self.inventory_database,
```

and format

File: src/databricks/labs/ucx/contexts/workflow_task.py [hunk 3]
```diff
@@ -50,13 +50,14 @@ def installation(self):
 
     @cached_property
     def jobs_crawler(self):
-        return JobsCrawler(self.workspace_client, self.sql_backend, self.inventory_database)
+        return JobsCrawler(self.workspace_client, self.sql_backend, self.inventory_catalog, self.inventory_database)
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/contexts/workflow_task.py:53]:
```suggestion
        return JobsCrawler(self.workspace_client, self.sql_backend, self.inventory_catalog, self.inventory_database, )
```

and format

File: src/databricks/labs/ucx/hive_metastore/locations.py [hunk 4]
```diff
@@ -300,8 +300,8 @@ def save_as_terraform_definitions_on_workspace(self, installation: Installation)
 
 
 class Mounts(CrawlerBase[Mount]):
-    def __init__(self, backend: SqlBackend, ws: WorkspaceClient, inventory_database: str):
-        super().__init__(backend, "hive_metastore", inventory_database, "mounts", Mount)
+    def __init__(self, backend: SqlBackend, ws: WorkspaceClient, inventory_catalog: str, inventory_database: str):
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/hive_metastore/locations.py:303]: Note that it is called `inventory_catalog` and `inventory_database` here where the others have just `catalog` and `schema`, maybe clean that up

File: src/databricks/labs/ucx/install.py [hunk 5]
```diff
@@ -227,12 +227,14 @@ def _is_testing(self):
     def _prompt_for_new_installation(self) -> WorkspaceConfig:
         logger.info("Please answer a couple of questions to configure Unity Catalog migration")
         default_database = "ucx"
+        default_catalog = "hive_metastore"
         # if a workspace is configured to use external hive metastore, the majority of the time that metastore will be
         # shared with other workspaces. we need to add the suffix to ensure uniqueness of the inventory database
         if self.policy_installer.has_ext_hms():
             default_database = f"ucx_{self.workspace_client.get_workspace_id()}"
+        inventory_catalog = self.prompts.question("Inventory Catalog", default=default_catalog, valid_regex=r"^\w+$")
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/install.py:235]: I do not think we want to do this. We still want the inventory to be in the `hive_metastore` (always), but need the drop in for the catalog to built the history table #2572

File: src/databricks/labs/ucx/mixins/fixtures.py [hunk 6]
```diff
@@ -970,6 +970,12 @@ def inventory_schema(make_schema):
     return make_schema(catalog_name="hive_metastore").name
 
 
+@pytest.fixture
+def inventory_catalog():
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/mixins/fixtures.py:974]: Is this needed? A test could just use `make_catalog` right?

File: src/databricks/labs/ucx/workspace_access/manager.py [hunk 7]
```diff
@@ -23,8 +23,10 @@ class PermissionManager(CrawlerBase[Permissions]):
 
     ERRORS_TO_IGNORE = ["FEATURE_DISABLED"]
 
-    def __init__(self, backend: SqlBackend, inventory_database: str, crawlers: list[AclSupport]):
-        super().__init__(backend, "hive_metastore", inventory_database, "permissions", Permissions)
+    def __init__(
+        self, backend: SqlBackend, inventory_catalog: str, inventory_database: str, crawlers: list[AclSupport]
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/workspace_access/manager.py:27]:
```suggestion
        self, backend: SqlBackend, inventory_catalog: str, inventory_database: str, crawlers: list[AclSupport],
```

File: tests/integration/aws/test_access.py [hunk 8]
```diff
@@ -105,6 +107,9 @@ def test_create_uber_instance_profile(ws, env_or_skip, make_random, make_cluster
     AWSResources(aws_cli_ctx.aws_profile()).delete_instance_profile(role_name, role_name)
 
 
+# TODO remove the disables and use ctx fixtures
```
Comment (JCZuurmond) [inline tests/integration/aws/test_access.py:110]: 👍

File: tests/integration/aws/test_credentials.py [hunk 9]
```diff
@@ -29,7 +29,7 @@ def inner(credentials: set[str], read_only=False) -> list[CredentialValidationRe
         )
 
         aws = AWSResources(env_or_skip("AWS_DEFAULT_PROFILE"))
-        location = ExternalLocations(ws, sql_backend, "inventory_schema")
+        location = ExternalLocations(ws, sql_backend, "hive_metastore", "inventory_schema")
```
Comment (JCZuurmond) [inline tests/integration/aws/test_credentials.py:32]: Not part of your changes: but would expect a dynamic schema or catalog here too

**Review by JCZuurmond — COMMENTED** [review]
Looking good, added some comments. Make sure you got all the `hive_metastore` references

File: src/databricks/labs/ucx/account/aggregate.py [hunk 10]
```diff
@@ -75,12 +75,13 @@ def _federated_ucx_query(self, query: str, table_name='objects') -> Iterable[tup
                         TableMigrationStatus(
                             src_schema=inventory_database,
                             src_table=table_name,
-                            dst_catalog='hive_metastore',
+                            dst_catalog=ctx.config.inventory_catalog,
```
Comment (nfx) [inline src/databricks/labs/ucx/account/aggregate.py:78]: revert, catalog will always be the same for this one. let's defer it to the time when we actually have our catalog.

File: src/databricks/labs/ucx/account/aggregate.py (hunk 1 above)
Comment (nfx) [reply src/databricks/labs/ucx/account/aggregate.py:84]: +1

File: src/databricks/labs/ucx/config.py [hunk 11]
```diff
@@ -11,6 +11,7 @@ class WorkspaceConfig:  # pylint: disable=too-many-instance-attributes
     __version__ = 2
 
     inventory_database: str
+    inventory_catalog: str = "hive_metastore"
```
Comment (nfx) [inline src/databricks/labs/ucx/config.py:14]: revert this change. we'll always have data in HMS (when we run run on existing workspaces)

**Review by nfx — CHANGES_REQUESTED** [review]
to align with @asnare

**Pull request closed without merging**